File: bbarena-model/src/main/java/net/sf/bbarena/model/event/game/NewTurnEvent.java
package net.sf.bbarena.model.event.game;

import net.sf.bbarena.model.Arena;
import net.sf.bbarena.model.TurnMarker;
import net.sf.bbarena.model.util.Concat;
import net.sf.bbarena.model.util.Pair;

public class NewTurnEvent extends GameEvent {

	private static final long serialVersionUID = -8405654897458550237L;

    private int _coachPos;
    private int _rerolls;
    private int _turn;
    private boolean _usedReroll;

    public NewTurnEvent() {
        super();
	}

    public int getCoachPos() {
        return _coachPos;
    }

    public void setCoachPos(int coachPos) {
        this._coachPos = coachPos;
    }

	@Override
	public void doEvent(Arena arena) {
        TurnMarker turnMarker = arena.getTurnMarkers().get(_coachPos);
        turnMarker.setStatus(TurnMarker.TurnStatus.PLAYING);
        _rerolls = turnMarker.getRerolls();
        _turn = turnMarker.getTurn();
        _usedReroll = turnMarker.isUsedReroll();

        turnMarker.setTurn(_turn + 1);
        turnMarker.setUsedReroll(false);
    }

	@Override
    protected void undoEvent(Arena arena) {
        TurnMarker turnMarker = arena.getTurnMarkers().get(_coachPos);
        turnMarker.setStatus(TurnMarker.TurnStatus.DONE);
        turnMarker.setUsedReroll(_usedReroll);
        turnMarker.setTurn(_turn);
        turnMarker.setRerolls(_rerolls);
    }

	@Override
	public String getString() {
        return Concat.buildLog(getClass(),
                new Pair("coach", _coachPos),
                new Pair("turn", _turn + 1));
    }

}

File: src/main/kotlin/bbarena/model/event/game/NewTurnEvent.kt
package bbarena.model.event.game

import bbarena.model.Arena
import bbarena.model.TurnMarker
import bbarena.model.util.Concat

class NewTurnEvent : GameEvent() {

    var coachPos: Int = 0
    private var rerolls: Int = 0
    private var turn: Int = 0
    private var usedReroll: Boolean = false

    override fun doEvent(arena: Arena) {
        val turnMarker = arena.turnMarkers[coachPos]
        turnMarker.status = TurnMarker.TurnStatus.PLAYING
        rerolls = turnMarker.rerolls
        turn = turnMarker.turn
        usedReroll = turnMarker.isUsedReroll

        turnMarker.turn = turn + 1
        turnMarker.isUsedReroll = false
    }

    override fun undoEvent(arena: Arena) {
        val turnMarker = arena.turnMarkers[coachPos]
        turnMarker.status = TurnMarker.TurnStatus.DONE
        turnMarker.isUsedReroll = usedReroll
        turnMarker.turn = turn
        turnMarker.rerolls = rerolls
    }

    override fun getString(): String {
        return Concat.buildLog(javaClass,
                Pair("coach", coachPos),
                Pair("turn", turn + 1))
    }

}
